Add unsecure hash function

jonas-lj opened the pull request:
Add a non-cryptographic hash function (Xxh3) for benchmarking. Closing #74

huitseeker commented on src/unsecure/mod.rs:
```diff
@@ -0,0 +1,8 @@
+/// This module contains an implementation of a negligible-cost (apart from some ocassional pk copying) trivial
+/// cryptographic signature scheme.
+///
+/// Warning: All schemes in this file are completely unsafe to use in production.
```
Besides the safety notice, I would add a directional comment indicating what we intend to use this for (benches). This would give a future auditor an easier time making heads or tails of this.

huitseeker reviewed (commented):
Same as #71: this needs at least couple of the simple remediation against accidental activation listed in https://gist.github.com/huitseeker/6f117d0157783b571c280c8462f4ab72 before merging.

kchalkias commented on src/unsecure/hash.rs:
```diff
@@ -0,0 +1,57 @@
+#[cfg(feature = "unsecure_schemes")]
+#[derive(Default)]
+pub struct XXH128Unsecure {
+    instance: twox_hash::xxh3::Hash128,
```
Wondering if we should on purpose use a scheme that outputs 32 bytes (so size  is compatible what most chains support): ie SuiAddress and Object derivation in Sui is truncating to 160 bits and having a shorter hash might cause issues.

jonas-lj replied:
I've considered the same thing. Xxhash currently [does not support 256 bit hashes](https://cyan4973.github.io/xxHash/), and to my knowledge, none of the fast non-cryptographic hash functions does. Bit in this case, we could simply concatenate two 128 bit or pad with zeroes, because we mostly care about avoiding collisions.

jonas-lj commented:
> Same as #71: this needs at least couple of the simple remediation against accidental activation listed in https://gist.github.com/huitseeker/6f117d0157783b571c280c8462f4ab72 before merging.

I've added two: There's a check that debug_assertions should be set, and I've also added a "secure" feature by default, and if both "secure" and "unsecure_schemes" are enabled, there'll be a compile error.

huitseeker reviewed (approved):
> I tried to add a compile error if both features were enables, but the clippy check failed because of that, because it runs with all features enabled.

I think it's 100% fair to make clippy run on default features and possibly make it have a distinct run on `--no-default-features --features "unsecure_schemes"`. clippy running with all features is just a CI setup shortcut with no deep meaning.

huitseeker commented on Cargo.toml:
```diff
@@ -54,8 +55,10 @@ name = "crypto"
 [features]
-default = []
+default = ["secure"]
+secure = []
```
Nit: we might want a comment here on how this is meant to control for accidental activations of `unsecure_schemes`.

kchalkias commented on src/unsecure/hash.rs:
```diff
@@ -0,0 +1,87 @@
+    }
+
+    fn finalize(self) -> Digest<typenum::U32> {
+        // Create a 32 byte digest consisting of two copies of a Xxh128 digest.
```
Thanks, that works! you could also pad it with zeros, but it's probably better this way to look like random 256bit string when printed.
